yksmp/src/lib.rs

```diff
@@ -0,0 +1,342 @@
+
+    fn read_u32(&mut self) -> u32 {
+        let d = u32::from_ne_bytes(self.data[self.offset..self.offset + 4].try_into().unwrap());
+        self.offset += 4;
+        d
+    }
+
+    fn read_i32(&mut self) -> i32 {
+        let d = i32::from_ne_bytes(self.data[self.offset..self.offset + 4].try_into().unwrap());
+        self.offset += 4;
+        d
+    }
+
+    fn read_u64(&mut self) -> u64 {
+        let d = u64::from_ne_bytes(self.data[self.offset..self.offset + 8].try_into().unwrap());
+        self.offset += 8;
+        d
+    }
+}
+
+struct Registers {
+    rax: usize,
+    rdx: usize,
+    rcx: usize,
+    rbx: usize,
+    rsi: usize,
+    rdi: usize,
+    rsp: usize,
+}
+
+impl Registers {
+    fn from_ptr(ptr: *const c_void) -> Registers {
+        Registers {
+            rax: Registers::read_from_stack(ptr, 0),
+            rdx: Registers::read_from_stack(ptr, 1),
+            rcx: Registers::read_from_stack(ptr, 2),
+            rbx: Registers::read_from_stack(ptr, 3),
+            rsi: Registers::read_from_stack(ptr, 4),
+            rdi: Registers::read_from_stack(ptr, 5),
+            rsp: Registers::read_from_stack(ptr, 6),
+        }
+    }
+
+    fn read_from_stack(rsp: *const c_void, off: isize) -> usize {
+        unsafe {
+            let ptr = rsp as *const usize;
+            ptr::read::<usize>(ptr.offset(off))
+        }
+    }
+
+    fn get(&self, id: u16) -> usize {
+        match id {
+            0 => self.rax,
+            1 => self.rdx,
+            2 => self.rcx,
+            3 => self.rbx,
+            4 => self.rsi,
+            5 => self.rdi,
+            6 => unreachable!("We currently have no way to access RBP of the previous stackframe."),
+            7 => self.rsp,
+            _ => unreachable!(),
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn __yk_stopgap(
+    addr: *const c_void,
+    size: usize,
+    retaddr: usize,
+    rsp: *const c_void,
+) {
+    // Restore saved registers from the stack.
+    let registers = Registers::from_ptr(rsp);
+
+    // Parse the stackmap.
+    let slice = unsafe { slice::from_raw_parts(addr as *mut u8, size) };
+    let map = StackMapParser::parse(slice).unwrap();
+    let locs = map.get(&retaddr.try_into().unwrap()).unwrap();
+
+    // Extract live values from the stackmap.
+    // FIXME: Until we have a stopgap interpreter we simply print the values.
+    for l in locs {
+        match l {
+            Location::Direct(reg, off) => {
+                // When using `llvm.experimental.deoptimize` then direct locations should always be
+                // in relation to RSP.
+                assert_eq!(*reg, 7);
+                // We need to add 2 bytes to the value of RSP to factor in the return address and
+                // the pushed RBP value of the previous frame.
+                let addr = registers.get(*reg) + (*off as usize) + 16;
+                let v = unsafe { ptr::read::<u8>(addr as *mut u8) };
+                println!("Direct: {}", v);
```

Comment:
I'm entirely allergic to this sort of `println`: let's make it a `todo()`.

Reply:
The reason I made this `print` instead of `todo` is so that we can test this using langtester until we have a stopgap interpreter. If I use todo I can only test the first value.

Reply:
But we're not testing this function at all in this PR are we?

Reply:
But we will in the very next PR tomorrow. I can remove it temporarily though I hardly see the point of doing so.

Reply:
I understand, but if we're not careful at some point we'll leave in a `print` that should be a `todo` and regret it. It's a good discipline for us to adhere to.

Reply:
Under normal circumstances I'd agree with you, but this is a bit of a special case. If we use `todo`s we won't be able to test this until we've implemented the stopgap interpreter which will be a while.

Reply:
In the next PR we can reconsider: but, and especially since this function isn't tested in this PR, these prints should be todos.

Reply:
Making @ltratt happy, one commit at a time: 659b4c1.

Reply:
It's so easy to make me happy :)